### PR TITLE
fix #35889 Wrong order of NOT NULL data type modifier and DEFAULT

### LINKED ISSUE #35889: Wrong order of NOT NULL data type modifier and DEFAULT.
**Describe the unexpected behaviour**

```
:) CREATE TEMPORARY TABLE test (x INT NOT NULL DEFAULT 1)

Syntax error: failed at position 52 ('DEFAULT') (line 3, col 22):

CREATE TEMPORARY TABLE test
(
    `x` INT NOT NULL DEFAULT 1
                     ^
)

Expected one of: COMMENT, CODEC, TTL, token, Comma, ClosingRoundBracket
```

But this query succeeded:

```
:) CREATE TEMPORARY TABLE test (x INT DEFAULT 1 NOT NULL)

CREATE TEMPORARY TABLE test
(
    `x` INT NOT NULL DEFAULT 1
)
```

### PR DESCRIPTION
**Pull request by attack204**
### Changelog category (leave one):

- Bug Fix (user-visible misbehaviour in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

[fix issue 35889](https://github.com/ClickHouse/ClickHouse/issues/35889)

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

The cause for the issue is that the original code only considered the situation that "DEFAULT" before "NOT NULL".
So I copy that code and then attach it to the "NOT NULL" code.

### CODE REVIEW

File: src/Parsers/ParserCreateQuery.h [hunk 1]
```diff
@@ -211,6 +211,21 @@ bool IParserColumnDeclaration<NameParser>::parseImpl(Pos & pos, ASTPtr & node, E
...
+            {
+                default_specifier = Poco::toUpper(std::string{pos_before_specifier->begin, pos_before_specifier->end});
+
+                /// should be followed by an expression
+                if (!expr_parser.parse(pos, default_expression, expected))
+                    return false;
+            }
+
+
+
```
Comment (ucasfl) [inline src/Parsers/ParserCreateQuery.h:228]: Stype check failed, more than two consecutive empty lines.

File: src/Parsers/ParserCreateQuery.h [hunk 2]
```diff
@@ -211,6 +211,18 @@ bool IParserColumnDeclaration<NameParser>::parseImpl(Pos & pos, ASTPtr & node, E
             if (!s_null.ignore(pos, expected))
                 return false;
             null_modifier.emplace(false);
+
+            pos_before_specifier = pos;
+            if (s_default.ignore(pos, expected) || s_materialized.ignore(pos, expected) ||
```
Comment (ucasfl) [inline src/Parsers/ParserCreateQuery.h:216]: Need to check does `default` has been parsed, current
 ```sql
CREATE TEMPORARY TABLE
(
    x INT DEFAULT 1 NOT NULL DEFAULT 2
)
``` 
seems also can be succussfully parsed.

Comment (attack204) [reply src/Parsers/ParserCreateQuery.h:216]: yeah. But it also can be parsed in Mysql XD.

Comment (ucasfl) [reply src/Parsers/ParserCreateQuery.h:216]: Hi, @evillique, what do you think?

**Comment by evillique** [comment]
@Mergifyio update

**Comment by devcrafter** [comment]
Fixed in #37337

**Pull request closed without merging**